File: python/ParkingLot.py
class ParkingLot:
    def __init__(self, motorcycleCapacity, carCapacity):
        self._motorcycleCapacity = motorcycleCapacity
        self._carCapacity = carCapacity
        self._listOfCar = []
        self._listOfMotorcycle = []

    # Search methods
    def searchCarPos(self, numberPlate):
        for i, car in enumerate(self._listOfCar):
            if car.getNumberPlate() == numberPlate:
                return i
        return -1

    def searchMotorcyclePos(self, numberPlate):
        for i, motorcycle in enumerate(self._listOfMotorcycle):
            if motorcycle.getNumberPlate() == numberPlate:
                return i
        return -1

    # Getters
    def getMotorcycleCapacity(self):
        return self._motorcycleCapacity

    def getCarCapacity(self):
        return self._carCapacity

    def getCars(self):
        return self._listOfCar

    def getMotorcycles(self):
        return self._listOfMotorcycle

    # Setters
    def setMotorcycleCapacity(self, capacity):
        self._motorcycleCapacity = capacity

    def setCarCapacity(self, capacity):
        self._carCapacity = capacity

    # Parking methods
    def addCar(self, car):
        if len(self._listOfCar) < self._carCapacity:
            self._listOfCar.append(car)
        else:
            print("There is no parking lot for", car.getNumberPlate())

    def addMotorcycle(self, motorcycle):
        if len(self._listOfMotorcycle) < self._motorcycleCapacity:
            self._listOfMotorcycle.append(motorcycle)
        else:
            print("There is no parking lot for", motorcycle.getNumberPlate())

    def startParkingTheCar(self, car):
        self.addCar(car)

    def startParkingTheMotorcycle(self, motorcycle):
        self.addMotorcycle(motorcycle)

    def finishParkingTheCar(self, car):
        pos = self.searchCarPos(car.getNumberPlate())
        if pos != -1:
            del self._listOfCar[pos]
        else:
            print("Number plate doesn't match")

    def finishParkingTheMotorcycle(self, motorcycle):
        pos = self.searchMotorcyclePos(motorcycle.getNumberPlate())
        if pos != -1:
            del self._listOfMotorcycle[pos]
        else:
            print("Number plate doesn't match")

    def print(self):
        print("Parking Lot")

        print("\nList of number plates of cars")
        for car in self._listOfCar:
            print(car.getNumberPlate())

        print("\nList of number plates of motorcycles")
        for motorcycle in self._listOfMotorcycle:
            print(motorcycle.getNumberPlate())

        print()

File: java/ParkingLot.java
import java.util.ArrayList;
import java.util.List;

public class ParkingLot {
    private int motorcycleCapacity;
    private int carCapacity;
    private List<Car> listOfCar;
    private List<Motorcycle> listOfMotorcycle;

    // Constructor
    public ParkingLot(int motorcycleCapacity, int carCapacity) {
        this.motorcycleCapacity = motorcycleCapacity;
        this.carCapacity = carCapacity;
        this.listOfCar = new ArrayList<>();
        this.listOfMotorcycle = new ArrayList<>();
    }

    // Destructor
    // No destructor needed in Java due to automatic garbage collection

    // Search methods
    private int searchCarPos(String numberPlate) {
        for (int i = 0; i < listOfCar.size(); i++) {
            if (listOfCar.get(i).getNumberPlate().equals(numberPlate)) {
                return i;
            }
        }
        return -1;
    }

    private int searchMotorcyclePos(String numberPlate) {
        for (int i = 0; i < listOfMotorcycle.size(); i++) {
            if (listOfMotorcycle.get(i).getNumberPlate().equals(numberPlate)) {
                return i;
            }
        }
        return -1;
    }

    // Getters
    public int getMotorcycleCapacity() {
        return motorcycleCapacity;
    }

    public int getCarCapacity() {
        return carCapacity;
    }

    public List<Car> getCars() {
        return listOfCar;
    }

    public List<Motorcycle> getMotorcycles() {
        return listOfMotorcycle;
    }

    // Setters
    public void setMotorcycleCapacity(int capacity) {
        motorcycleCapacity = capacity;
    }

    public void setCarCapacity(int capacity) {
        carCapacity = capacity;
    }

    // Parking methods
    public void addCar(Car car) {
        if (listOfCar.size() < carCapacity) {
            listOfCar.add(car);
        } else {
            System.out.println("There is no parking lot for " + car.getNumberPlate());
        }
    }

    public void addMotorcycle(Motorcycle motorcycle) {
        if (listOfMotorcycle.size() < motorcycleCapacity) {
            listOfMotorcycle.add(motorcycle);
        } else {
            System.out.println("There is no parking lot for " + motorcycle.getNumberPlate());
        }
    }

    public void startParkingTheCar(Car car) {
        addCar(car);
    }

    public void startParkingTheMotorcycle(Motorcycle motorcycle) {
        addMotorcycle(motorcycle);
    }

    public void finishParkingTheCar(Car car) {
        int pos = searchCarPos(car.getNumberPlate());
        if (pos != -1) {
            listOfCar.remove(pos);
        } else {
            System.out.println("Number plate doesn't match");
        }
    }

    public void finishParkingTheMotorcycle(Motorcycle motorcycle) {
        int pos = searchMotorcyclePos(motorcycle.getNumberPlate());
        if (pos != -1) {
            listOfMotorcycle.remove(pos);
        } else {
            System.out.println("Number plate doesn't match");
        }
    }

    public void print() {
        System.out.println("Parking Lot");

        System.out.println("\nList of number plates of cars");
        for (Car car : listOfCar) {
            System.out.println(car.getNumberPlate());
        }

        System.out.println("\nList of number plates of motorcycles");
        for (Motorcycle motorcycle : listOfMotorcycle) {
            System.out.println(motorcycle.getNumberPlate());
        }

        System.out.println();
    }
}
